File: matpy/matrix.py
# ---------------------------------------------------
# Name: Ahmad Amin, Yash Bhandari & Nibras Alam
# ID: 1623338, 1616539 & 1617818
# CMPUT 275, Winter 2020
#
# Final Project
#
# Consulted Python Documentation
# ---------------------------------------------------
from . import _Matrix  # c++ Matrix implementation
import numpy as np

# function decorator to wrap c++ matrices in a python Matrix
def wrap_cpp_matrix(func):
    def wrapped(*args, **kwargs):
        return Matrix(cpp_matrix=func(*args, **kwargs))

    return wrapped


class Matrix:
    def __init__(self, matrix_values=[[0]], cpp_matrix=None):
        """
        Args:
            matrix_values: 2D list or Matrix from which to initialize values
        """
        if cpp_matrix:
            self._mat = cpp_matrix
        else:
            self._mat = _Matrix(matrix_values)

    def __getitem__(self, indices):
        """
        Args:
            indices: a tuple (i, j) representing indices of a matrix entry
        Returns:
            the i,j-th entry of the matrix
        """
        return self._mat.get(indices[0], indices[1])

    @wrap_cpp_matrix
    def __add__(self, other: 'Matrix') -> 'Matrix':
        """
        Args:
            other: another matrix of the same dimensions
        Returns:
            The matrix sum of self and other
        """
        return self._mat.add(other._mat)

    @wrap_cpp_matrix
    def __matmul__(self, other: 'Matrix') -> 'Matrix':
        """
        Distinct from __mul__ to comply with PEP 465
        Args:
            other: another matrix of the same dimensions
        Returns:
            The matrix product of self and other
        """
        return self._mat.mult(other._mat)

    @wrap_cpp_matrix
    def __mul__(self, scalar: int) -> 'Matrix':
        """
        Args:
            scalar: an integer
        Returns:
            The matrix scalar product
        """
        return self._mat.scale(scalar)

    @wrap_cpp_matrix
    def __rmul__(self, scalar: int) -> 'Matrix':
        """
        Args:
            scalar: an integer
        Returns:
            The matrix scalar product
        """
        return self._mat.scale(scalar)

    @wrap_cpp_matrix
    def __pow__(self, pow: int) -> 'Matrix':
        """
        Args:
            pow: a non negative integer
        Returns:
            This matrix raised to the given power
        """
        return self._mat.exp(pow)

    def __eq__(self, other):
        """
        Args:
            other: Another matrix or numpy array
        Returns:
            True if the two matrices are equal
        """
        if isinstance(other, np.ndarray):
            return self.__equals_numpy_array(other)
        if not isinstance(other, Matrix):
            return False
        return self._mat.equals(other._mat)

    def __equals_numpy_array(self, np_array):
        """
        Compares with numpy array
        """
        if self.dim != np_array.shape:
            return False

        for i in range(self.dim[0]):
            for j in range(self.dim[1]):
                if self[i,j] != np_array[i,j]:
                    return False
        return True
    
    def __str__(self):
        return self._mat.str()

    def __repr__(self):
        return self.__str__()

    @property
    def dim(self):
        """
        Returns the dimension of the matrix as a tuple
        """
        return self._mat.dim()

    @classmethod
    @wrap_cpp_matrix
    def id(cls, n):
        """
        Constructs an n x n identity matrix
        """
        return _Matrix.id(n)

    @classmethod
    @wrap_cpp_matrix
    def fill(cls, m, n, fill=0):
        """
        Constructs an m x n matrix with all entries set to the fill parameter
        """
        return _Matrix(m, n, fill)

    @classmethod
    @wrap_cpp_matrix
    def naive_mult(cls, A, B):
        """
        Uses a naive matrix multiplication algorithm to compute A*B
        """
        return _Matrix.naive_mult(A._mat, B._mat)

    @classmethod
    @wrap_cpp_matrix
    def strassen_mult(cls, A, B):
        """
        Uses the strassen algorithm to compute the product of A and B
        """
        return _Matrix.strassen_mult(A._mat, B._mat)

File: matpy/src/matrix.cpp
/*
# Name: Ahmad Amin, Yash Bhandari & Nibras Alam
# ID: 1623338, 1616539 & 1617818
# CMPUT 275, Winter 2020
#
# Final Project
#
# Consulted C++ Documentation;
*/
#include <pybind11/pybind11.h>
#include <pybind11/stl.h>
#include <exception>
#include <iostream>
#include "matrix.h"
#include "matrix_ops.h"

namespace py = pybind11;

bool Matrix::wellFormed(std::vector<std::vector<int>> matrixValues)
{
    size_t n = matrixValues[0].size();
    for (auto row : matrixValues)
        if (row.size() != n)
            return false;
    return true;
}

Matrix *Matrix::id(size_t n)
{
    Matrix *idMat = new Matrix(n, n);
    for (size_t i = 0; i < n; i++)
    {
        idMat->set(i, i, 1);
    }
    return idMat;
}

Matrix::Matrix(size_t m, size_t n, int fill)
    : m(m), n(n)
{
    mat = new int[m * n];
    for (size_t i = 0; i < m * n; i++)
        mat[i] = fill;
}

Matrix::Matrix(std::vector<std::vector<int>> matrixValues)
{
    if (!wellFormed(matrixValues))
        throw std::invalid_argument("Matrix had rows of different lengths");

    m = matrixValues.size();
    n = matrixValues[0].size();
    mat = new int[m * n];
    for (size_t i = 0; i < m; i++)
    {
        for (size_t j = 0; j < n; j++)
        {
            mat[i * n + j] = matrixValues[i][j];
        }
    }
}

Matrix::~Matrix()
{
    delete[] mat;
}

std::string Matrix::str() const
{
    std::string result = std::to_string(m) + 'x' + std::to_string(n) + " Matrix:";
    for (size_t i = 0; i < m; i++)
    {
        result += '\n';
        for (size_t j = 0; j < n; j++)
        {
            result += std::to_string(mat[i * n + j]) + ' ';
        }
    }
    return result;
}

int Matrix::get(size_t i, size_t j) const
{
    return mat[i * n + j];
}

void Matrix::set(size_t i, size_t j, int val)
{
    mat[i * n + j] = val;
}

std::tuple<size_t, size_t> Matrix::dim() const
{
    return std::tuple<size_t, size_t>(m, n);
}

Matrix *Matrix::add(Matrix const *other) const
{
    if (dim() != other->dim())
    {
        std::string dim1 = formatDim(other->dim());
        std::string dim2 = formatDim(dim());
        throw std::invalid_argument("Matrix of dimension " + dim1 + " cannot \
                            be multiplied with matrix of dimension " +
                                    dim2);
    }

    Matrix *sum = new Matrix(m, n);
    for (size_t i = 0; i < m; i++)
    {
        for (size_t j = 0; j < n; j++)
        {
            int entry = other->get(i, j) + get(i, j);
            sum->set(i, j, entry);
        }
    }
    return sum;
}

Matrix *Matrix::mult(Matrix const *other) const
{
    if (numCols() != other->numRows())
    {
        std::string dim1 = formatDim(other->dim());
        std::string dim2 = formatDim(dim());
        throw std::invalid_argument("Matrix of dimension " + dim1 + " cannot \
                            be multiplied with matrix of dimension " +
                                    dim2);
    }

    auto multAlgorithm = optimalMultAlgorithm(this, other);
    return multAlgorithm(this, other);
}

Matrix *Matrix::scale(int scalar) const
{
    Matrix *result = new Matrix(numRows(), numCols());
    for (size_t i = 0; i < numRows(); i++)
    {
        for (size_t j = 0; j < numCols(); j++)
        {
            int scaled = scalar * get(i, j);
            result->set(i, j, scaled);
        }
    }
    return result;
}

bool Matrix::equals(Matrix const *other) const
{
    if (dim() != other->dim())
    {
        return false;
    }

    for (size_t i = 0; i < m; i++)
    {
        for (size_t j = 0; j < n; j++)
        {
            if (get(i, j) != other->get(i, j))
            {
                return false;
            }
        }
    }
    return true;
}

Matrix *Matrix::exp(unsigned int x)
{
    if (numRows() != numCols())
    {
        throw std::invalid_argument("Cannot exponentiate matrix of dimension " + formatDim(dim()) + ". Matrix must be square.");
    }
    // either return identity matrix or prepare for calculation
    Matrix *result;
    if (x == 0)
    {
        result = id(numRows());
    }
    else
    {
        // copy matrix to result matrix (A^1)
        result = new Matrix(numRows(), numCols());
        for (size_t i = 0; i < numRows(); i++)
        {
            for (size_t j = 0; j < numCols(); j++)
            {
                result->set(i, j, get(i, j));
            }
        }

        // repeatedly multiply matrix to result matrix(A^x)
        for (unsigned int i = 1; i < x; i++)
        {
            result = mult(result);
        }
    }

    return result;
}

size_t Matrix::numRows() const
{
    return m;
}

size_t Matrix::numCols() const
{
    return n;
}

std::string Matrix::formatDim(std::tuple<size_t, size_t> dimension)
{
    std::string dim1 = std::to_string(std::get<0>(dimension));
    std::string dim2 = std::to_string(std::get<1>(dimension));
    return "(" + dim1 + ", " + dim2 + ")";
}
